### PR TITLE
feat(new_metrics): support to find multiple nth elements of a sequence container at a time based on nth_element() of STL

### PR DESCRIPTION
**Pull request by empiredan**
For details please see https://github.com/apache/incubator-pegasus/issues/974. The results of benchmark are also shown here.

# Benchmark
##  Hardware configurations
``` 
cpu: Intel Xeon Processor (Cascadelake)
cores: 4
memory: 32GB
```

## Software configurations
```
architecture: x86_64
gcc version: 7.3.1 20180303 (Red Hat 7.3.1-5)
```

## Performance
In `perf_counter_number_percentile_atomic` the size of sampled window is 5000 (MAX_QUEUE_LENGTH), thus this size is also adopted for the benchmark. And an operation is defined as a computation to find P50, P90, P95, P99 and P999 over the sampled window.

The dataset for the benchmark is generated randomly with range size 5 (see the code of bench for details). The following table lists the comparison between the execution time of both solutions:

| Number of operations (windows size: 5000)  | Solution based on `perf_counter_number_percentile_atomic` | Solution based on nth_element() |
| :-------------: | :-------------: | :-------------: |
| 1 | 0.728 ms | 0.384 ms |
| 10 |  7.127 ms | 3.256 ms |
| 100 |  72.030 ms |  33.378 ms |
| 1000 |  0.725 s |  0.331 s |
| 10000 | 7.185 s |  3.281 ms |
| 100000 | 72.358 s |  33.085 s |

### CODE REVIEW

File: include/dsn/utility/nth_element.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+#include <utility>
+#include <vector>
+
+#include <dsn/c/api_utilities.h>
+#include <dsn/dist/fmt_logging.h>
+#include <dsn/utility/ports.h>
+
+namespace dsn {
+
+// The finder helps to find multiple nth elements of a sequence container (e.g., std::vector)
```
Comment (acelyc111) [inline include/dsn/utility/nth_element.h:34]:
```suggestion
// The finder helps to find multiple nth elements of a sequence container (e.g. std::vector)
```
```suggestion
// The finder helps to find multiple nth elements of a sequence container (e.g., std::vector)
```

Comment (acelyc111) [inline include/dsn/utility/nth_element.h:34]:
```suggestion
// The finder helps to find multiple nth elements of a sequence container (e.g. std::vector)
```
```suggestion
// The finder helps to find multiple nth elements of a sequence container (e.g., std::vector)
```

File: include/dsn/utility/nth_element.h [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+template <typename T,
+          typename Compare = std::less<T>,
+          typename ContainerType = typename std::vector<T>>
+class stl_nth_element_finder
+{
+public:
+    using value_type = T;
+    using container_type = ContainerType;
+    using size_type = typename container_type::size_type;
+    using nth_container_type = std::vector<size_type>;
```
Comment (acelyc111) [inline include/dsn/utility/nth_element.h:45]: How about use std::set instead of std::vector? then you can avoid to sort indexes, and `set_nths()` and `nths()` are consistency in semantics.

File: include/dsn/utility/nth_element.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    // It is allowed that there are identical indexes in `nths` list, since it's also a
+    // sequence container.
+    void set_nths(const nth_container_type &nths)
+    {
+        _nths = nths;
+        std::sort(_nths.begin(), _nths.end());
+        _elements.assign(_nths.size(), value_type{});
+    }
+
+    const nth_container_type &nths() { return _nths; }
```
Comment (acelyc111) [inline include/dsn/utility/nth_element.h:65]:
```suggestion
    const nth_container_type &nths() const { return _nths; }
```

File: include/dsn/utility/nth_element.h [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    // in this function. Later the elements returned by `elements()` will be in the order
+    // corresponding to the sorted nth indexes, which is also returned by `nths()`.
+    //
+    // It is allowed that there are identical indexes in `nths` list, since it's also a
+    // sequence container.
+    void set_nths(const nth_container_type &nths)
+    {
+        _nths = nths;
+        std::sort(_nths.begin(), _nths.end());
+        _elements.assign(_nths.size(), value_type{});
```
Comment (acelyc111) [inline include/dsn/utility/nth_element.h:62]: Use `reserve` is OK? no need to assign values here.

File: include/dsn/utility/nth_element.h [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            // Identical nth indexes should be processed.
+            for (++i; i < _nths.size() && _nths[i] == _nths[i - 1]; ++i) {
+                _elements[i] = *nth_iter;
+            }
+
+            first = nth_iter + 1;
+        }
+    }
+
+    const container_type &elements() { return _elements; }
```
Comment (acelyc111) [inline include/dsn/utility/nth_element.h:92]:
```suggestion
    const container_type &elements() const { return _elements; }
```

File: include/dsn/utility/nth_element.h [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    operator()(RandomAccessIterator begin, RandomAccessIterator first, RandomAccessIterator last)
+    {
+        for (size_type i = 0; i < _nths.size();) {
+            auto nth_iter = begin + _nths[i];
+            dassert_f(nth_iter >= first && nth_iter < last, "Invalid iterators for nth_element()");
+            std::nth_element(first, nth_iter, last, _comp);
+            _elements[i] = *nth_iter;
+
+            // Identical nth indexes should be processed.
+            for (++i; i < _nths.size() && _nths[i] == _nths[i - 1]; ++i) {
```
Comment (acelyc111) [inline include/dsn/utility/nth_element.h:84]: Is there any use case we have to support `_nths[i] == _nths[i - 1]`?

File: src/perf_counter/perf_counter_atomic.h [hunk 7]
```diff
@@ -197,7 +197,8 @@ class perf_counter_number_percentile_atomic : public perf_counter
                                           const char *section,
                                           const char *name,
                                           dsn_perf_counter_type_t type,
-                                          const char *dsptr)
+                                          const char *dsptr,
+                                          bool use_timer = true)
```
Comment (acelyc111) [inline src/perf_counter/perf_counter_atomic.h:201]: Is this modification related to this PR ?

**Comment by acelyc111** [comment]
Could you please add benchmark result in this PR describtion.

File: src/utils/test/nth_element_bench/nth_element_bench.cpp [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+#include <dsn/utility/nth_element.h>
+#include <dsn/utility/smart_pointers.h>
+#include <dsn/utility/string_conv.h>
+#include <dsn/utility/strings.h>
+
+#include "utils/test/nth_element_utils.h"
+
+void print_usage(const char *cmd)
+{
+    fmt::print(stderr, "USAGE: {} <num_operations> <array_size> [nths]\n", cmd);
```
Comment (acelyc111) [inline src/utils/test/nth_element_bench/nth_element_bench.cpp:40]: Is it nessesary to output usage into stderr?

File: include/dsn/utility/nth_element.h [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+public:
+    using value_type = T;
+    using container_type = std::vector<value_type>;
+    using size_type = typename container_type::size_type;
+    using nth_container_type = std::vector<size_type>;
+
+    stl_nth_element_finder(const Compare &comp = Compare()) : _nths(), _elements(), _comp(comp) {}
+
+    // Set with specified nth indexes. An nth index is typically an index of the sequence
+    // container (std::vector). This method allows nth indexes to be updated dynamically.
```
Comment (acelyc111) [inline include/dsn/utility/nth_element.h:50]:
```suggestion
    // container (e.g. std::vector). This method allows nth indexes to be updated dynamically.
```

**Review by acelyc111 — APPROVED** [review]
LGTM

File: src/perf_counter/perf_counter_atomic.h [hunk 10]
```diff
@@ -214,12 +219,17 @@ class perf_counter_number_percentile_atomic : public perf_counter
             "pegasus_perf_counter_number_percentile_atomic counters");
         _timer.reset(new boost::asio::deadline_timer(tools::shared_io_service::instance().ios));
         _timer->expires_from_now(
-            boost::posix_time::seconds(rand() % _counter_computation_interval_seconds + 1));
+            boost::posix_time::seconds(::rand() % _counter_computation_interval_seconds + 1));
```
Comment (levy5307) [inline src/perf_counter/perf_counter_atomic.h:222]: Why did you add the prefix :: ?

File: src/perf_counter/perf_counter_atomic.h [hunk 11]
```diff
@@ -214,12 +219,17 @@ class perf_counter_number_percentile_atomic : public perf_counter
...
         _timer->async_wait(std::bind(
             &perf_counter_number_percentile_atomic::on_timer, this, _timer, std::placeholders::_1));
     }
 
-    ~perf_counter_number_percentile_atomic(void) { _timer->cancel(); }
+    ~perf_counter_number_percentile_atomic(void)
+    {
+        if (_timer) {
```
Comment (levy5307) [inline src/perf_counter/perf_counter_atomic.h:229]: for the if statement, it`s better using `if (_timer != nullptr)` if the type of _time is a pointer

File: src/perf_counter/perf_counter_atomic.h (hunk 10 above)
Comment (empiredan) [reply src/perf_counter/perf_counter_atomic.h:222]: > Why did you add the prefix :: ?

Without prefixal `::`, rDSN will be failed to be compiled, showing error message: **error: expected primary-expression before ‘(’ token**. Since it's been included by other source files such as `src/utils/test/nth_element_utils.h` and `src/utils/test/nth_element_bench/nth_element_bench.cpp`, `rand()` may has conflicted with some functions of other header files. Therefore, I've added a prefixal `::` to resolve.

File: src/perf_counter/perf_counter_atomic.h (hunk 11 above)
Comment (empiredan) [reply src/perf_counter/perf_counter_atomic.h:229]: > for the if statement, it`s better using `if (_timer != nullptr)` if the type of _time is a pointer

Since `_timer` is a `std::shared_ptr`, we can use `operator bool `(https://en.cppreference.com/w/cpp/memory/shared_ptr/operator_bool) to check if a std::shared_ptr has stored a valid pointer, shall we ?

Comment (levy5307) [reply src/perf_counter/perf_counter_atomic.h:229]: OK

**Review by levy5307 — APPROVED** [review]
LGTM
